### DISCUSSION TITLE
Is there a way to bypass the network call when the caller and rqlite are on the same host?

### DISCUSSION THREAD

**Discussion by kampilan** [rqlite/rqlite#1983] [General]
Is there a way to submit SQL statements directly to rqlite when the caller and rqlite are on the same host? Does rqlite support Unix domain sockets maybe? Or some other IPC protocol?

One other unrelated question. Is rqlite multi-master? Can I write to any node?

Thanks so much
Jim

**Comment by otoolep** [comment]
There is no Unix socket support: https://github.com/rqlite/rqlite/issues/1737 - see my response on that ticket. I'm not convinced it's worth adding, but I could be convinced if someone presented a compelling use case to me. However HTTP is generally functional as an IPC protocol, though I agree it's not as fancy as others.

rqlite is not-multi-master. It's a Raft-based system, which means it has single **Leader**. However you can write **to any node in the cluster**, rqlite will handle it all for you, you don't need to worry about sending writes to the Leader. Send your writes to any node in the cluster, it'll all work.

https://rqlite.io/docs/faq/#can-any-node-execute-a-write-request-and-have-the-system-synchronize-it-all

There is extensive documentation available at https://rqlite.io/docs/

**Comment by kampilan** [comment]
Sorry I only lightly scanned the discussions before posting. Mutli-master was my ignorant way of describing the ability to write to any node functionality.

Would your opinion change any if I said I was going to use Queued Writes? My use case doesn't care that much about write latency or the risk of data loss. What it does care about is returning to the caller as fast as possible. Maybe I am obsessing over the perceived performance of UDS over TCP. I'll just try rqlite and see how it performs.

Thanks for the quick reply. 
Jim

**Comment by otoolep** [comment]
I doubt the performance difference will matter. You'll saturate the disk long before you hit any UDS/HTTP bottleneck. See this part my talk on Raft to understand why: https://youtu.be/8XbxQ1Epi5w?t=2050

**Comment by otoolep** [comment]
Queued Writes will make a big difference however, because the fsyncs to the Raft log will be amortized over much more data. But again, the disk will become your bottleneck long before your network does.

**Comment by kampilan** [comment]
Agreed. Thank you for your time.

**Comment by otoolep** [comment]
>My use case doesn't care that much about write latency or the risk of data loss.

Queued Writes sounds like exactly like what you need.

**Reply by kampilan** [reply]
As soon as I saw it, I realized it might be a perfect fit. It's for a structured logging sink. If the writes are a little bit behind, no problem. While not ideal to lose of some logging records worth it to get high availability.

**Reply by otoolep** [reply]
Yes, that sounds like a good use case.

**Reply by otoolep** [reply]
You might be interested in this blog post: https://philipotoole.com/building-a-highly-available-search-engine-using-sqlite/

In it I talk about indexing Apache web server logs.

**Comment by otoolep** [comment]
https://rqlite.io/docs/api/queued-writes/
